// lorejs/games/lemegeton.js
const LemegetonPlugin = {
  id: "lemegeton_plugin",
  name: "Lemegeton Reading Plugin",
  commands: [
    {
      name: "read",
      aliases: ["consult"],
      help: "Read the Lemegeton grimoire. Use: 'read lemegeton', 'read page [number]', 'read next page', etc.",
      fn: (args, engine) => {
        const itemName = args.join(" ").toLowerCase();
        const lemegeton = engine.world.items.get('lemegeton');
        
        if (!engine.state.inventory.includes('lemegeton')) {
          engine.printLine("You are not carrying the Lemegeton.");
          return true;
        }

        const book = lemegeton;
        if (!book.state) {
          book.state = { lastPage: 0, length: 50 };
        }

        // Handle relative pages
        if (itemName.includes("first page")) {
          return readPage(1, engine, book);
        } else if (itemName.includes("last page")) {
          return readPage(book.state.length, engine, book);
        } else if (itemName.includes("next page")) {
          return readPage(book.state.lastPage + 1, engine, book);
        } else if (itemName.includes("previous page")) {
          return readPage(book.state.lastPage - 1, engine, book);
        }

        // Handle specific page number
        const pageMatch = itemName.match(/page (\d+)/);
        if (pageMatch) {
          return readPage(parseInt(pageMatch[1]), engine, book);
        }

        // Handle "read lemegeton" or just "read"
        if (args.length === 0 || itemName.includes("lemegeton") || itemName.includes("grimoire")) {
          const randomPage = Math.floor(Math.random() * book.state.length) + 1;
          engine.printLine(`You flip the pages randomly and arrive at page ${randomPage}:`);
          return readPage(randomPage, engine, book);
        }

        engine.printLine("Read what?");
        return true;
      }
    }
  ]
};

function readPage(n, engine, book) {
  if (n < 1) {
    engine.printLine("The page numbering begins with 1.");
    return true;
  }
  if (n > book.state.length) {
    engine.printLine(`There are only ${book.state.length} pages in the book.`);
    return true;
  }

  book.state.lastPage = n;

  if (n === 47) {
    engine.printLine("Your eyes burn; your ears ring. Beneath your gaze, the dreadful sigils writhe, reminding you of that which lies outside the edges of the universe...");
    engine.printLine("{{bold}}{{red}}You have lost your remaining sanity.{{font_reset}}");
    setTimeout(() => {
        engine.gameOver();
    }, 2000);
    return true;
  }

  const contents = {
    1: "The first page contains the seal of Bael, the first king of the East.",
    2: "Page 2 details the summoning of Agares, a duke who can make those who run stand still.",
    10: "A diagram of the brass vessel used to contain the 72 spirits.",
    13: "Faded ink describes the properties of the Ring of Solomonis.",
    25: "A list of the planetary hours and their corresponding spirits.",
    50: "The final page is covered in cryptic warnings about the Tree of Forbidden Knowledge."
  };

  if (contents[n]) {
    engine.printLine(`You read:{{n}}{{is}}'${contents[n]}'{{fr}}`);
  } else {
    engine.printLine(`Page ${n} appears to be blank.`);
  }
  return true;
}

const LemegetonNovel = {
  title: 'Lemegeton: The Journey to Apotheosis',
  startRoom: 'atrium',
  onLoad: (engine) => {
    engine.loadPlugin(LemegetonPlugin);
  },
  rooms: [
    {
      id: 'atrium',
      name: '{{bold}}Atrium{{font_reset}}',
      description: 'A ritual atrium filled with humming glyphs. The Lemegeton grimoire lies on a pedestal here. Try commands like {{yellow}}help{{color_reset}}, {{yellow}}go north{{color_reset}}, {{yellow}}take lemegeton{{color_reset}}, and {{yellow}}read lemegeton{{color_reset}}.',
      exits: {
        north: 'sanctum',
        south: 'mothers_cottage'
      },
      items: ['sigil', 'lemegeton'],
      onEnter: (engine) => {
        if (!engine.plugins.has('lemegeton_plugin')) {
          engine.loadPlugin(LemegetonPlugin);
        }
        if (!engine.state.flags.metDemons) {
            setTimeout(() => {
                engine.printLine("Astaroth: \"Welcome, seeker. You look lost in these halls of knowledge.\"");
                engine.printLine("Belial: \"Knowledge is a heavy burden, isn't it? Better to just enjoy the view.\"");
                engine.state.flags.metDemons = true;
            }, 500);
        }
      },
      characters: ['astaroth', 'belial'],
    },
    {
      id: 'sanctum',
      name: '{{bold}}Sanctum{{font_reset}}',
      description: 'A quiet sanctum where idle effects continue between actions.',
      exits: {
        south: 'atrium'
      }
    }
  ],
  characters: [
    {
      id: 'astaroth',
      name: 'Astaroth',
      description: 'A great and powerful duke, who appears as an angel with a dragon-like breath.'
    },
    {
      id: 'belial',
      name: 'Belial',
      description: 'A king of the abyss, created next after Lucifer.'
    }
  ],
  items: [
    {
      id: 'sigil',
      name: 'metal sigil seal',
      takeable: true,
      description: 'A metal sigil that pulses whenever action animations trigger.',
      use: (engine) => {
        engine.printLine('The sigil resonates with your command flow.');
        return true;
      }
    },
    {
      id: 'lemegeton',
      name: 'Lemegeton grimoire',
      aliases: ['book', 'grimoire', 'lemegeton'],
      takeable: true,
      description: 'An ancient grimoire containing forbidden knowledge of spirit summoning.',
      use: (engine) => {
        // Using the book triggers a random read
        const book = engine.world.items.get('lemegeton');
        const randomPage = Math.floor(Math.random() * (book.state?.length || 50)) + 1;
        engine.printLine(`You open the Lemegeton:`);
        return readPage(randomPage, engine, book);
      }
    }
  ]
};

if (typeof module !== 'undefined' && module.exports) {
  module.exports = LemegetonNovel;
} else if (typeof window !== 'undefined') {
  window.LemegetonNovel = LemegetonNovel;
}
